mujoco_warp/_src/derivative.py

```diff
@@ -210,6 +211,496 @@ def _qderiv_tendon_damping(
+
+  D[0, 0] -= dfx_coef * proj_num
+  D[1, 1] -= dfy_coef * proj_num
+  D[2, 2] -= dfz_coef * proj_num
+
+  return D * df_denom
+
+
+@wp.func
+def _magnus_force_deriv(
+  lvel: wp.spatial_vector,
+  fluid_density: float,
+  size: wp.vec3,
+  magnus_lift_coef: float
+) -> wp.mat66:
+  volume = wp.static(4.0 / 3.0 * wp.pi) * size[0] * size[1] * size[2]
+  magnus_coef = magnus_lift_coef * fluid_density * volume
+
+  lin_vel = wp.spatial_bottom(lvel)
+  ang_vel = wp.spatial_top(lvel)
+
+  lin_vel_scaled = lin_vel * magnus_coef
+  ang_vel_scaled = ang_vel * magnus_coef
+
+  D_ang = _cross_deriv_a(ang_vel_scaled, lin_vel)
+  D_lin = _cross_deriv_b(ang_vel, lin_vel_scaled)
+
+  B = wp.mat66(0.0)
+  B = _add_to_quadrant(B, D_ang, 1, 0)
+  B = _add_to_quadrant(B, D_lin, 1, 1)
+
+  return B
+
+
+@wp.kernel
+def _qderiv_ellipsoid_fluid(
+  # Model:
+  opt_density: wp.array(dtype=float),
+  opt_viscosity: wp.array(dtype=float),
+  opt_wind: wp.array(dtype=wp.vec3),
+  opt_timestep: wp.array(dtype=float),
+  opt_is_sparse: bool,
+  body_rootid: wp.array(dtype=int),
+  body_geomnum: wp.array(dtype=int),
+  body_geomadr: wp.array(dtype=int),
+  body_fluid_ellipsoid: wp.array(dtype=bool),
+  geom_type: wp.array(dtype=int),
+  geom_size: wp.array2d(dtype=wp.vec3),
+  geom_fluid: wp.array2d(dtype=float),
+  dof_bodyid: wp.array(dtype=int),
+  # Data in:
+  xipos_in: wp.array2d(dtype=wp.vec3),
+  ximat_in: wp.array2d(dtype=wp.mat33),
+  geom_xpos_in: wp.array2d(dtype=wp.vec3),
+  geom_xmat_in: wp.array2d(dtype=wp.mat33),
+  subtree_com_in: wp.array2d(dtype=wp.vec3),
+  cvel_in: wp.array2d(dtype=wp.spatial_vector),
+  cdof_in: wp.array2d(dtype=wp.spatial_vector),
+  # In:
+  qMi: wp.array(dtype=int),
+  qMj: wp.array(dtype=int),
+  # Out:
+  qDeriv_out: wp.array3d(dtype=float),
+):
+  worldid, elemid = wp.tid()
+  dofiid = qMi[elemid]
+  dofjid = qMj[elemid]
+
+  bodyid_i = dof_bodyid[dofiid]
+  bodyid_j = dof_bodyid[dofjid]
+
+  if bodyid_i == 0 or not body_fluid_ellipsoid[bodyid_i]:
+    return
+
+  if bodyid_i != bodyid_j:
+    return
+
+  bodyid = bodyid_i
+  wind = opt_wind[worldid % opt_wind.shape[0]]
+  density = opt_density[worldid % opt_density.shape[0]]
+  viscosity = opt_viscosity[worldid % opt_viscosity.shape[0]]
+  timestep = opt_timestep[worldid % opt_timestep.shape[0]]
+
+  xipos = xipos_in[worldid, bodyid]
+  rot = ximat_in[worldid, bodyid]
+  rotT = wp.transpose(rot)
+  cvel = cvel_in[worldid, bodyid]
+  ang_global = wp.spatial_top(cvel)
+  lin_global = wp.spatial_bottom(cvel)
+  subtree_root = subtree_com_in[worldid, body_rootid[bodyid]]
+  lin_com = lin_global - wp.cross(xipos - subtree_root, ang_global)
+
+  qderiv_contrib = float(0.0)
+
+  start = body_geomadr[bodyid]
+  count = body_geomnum[bodyid]
+
+  for i in range(count):
+    geomid = start + i
+    coef = geom_fluid[geomid, 0]
+    if coef <= 0.0:
+      continue
+
+    size = geom_size[worldid % geom_size.shape[0], geomid]
+    semiaxes = _geom_semiaxes_deriv(size, geom_type[geomid])
```

Comment:
lets import `_geom_semiaxes` from `passive.py`. additionally let's change the name (in passive.py) to `geom_semiaxes`